### PR TITLE
Added PAT metric

### PR DESCRIPTION
**Pull request by mohan1914**
Add PAT metric for panoptic tracking evaluation.

### CODE REVIEW

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 1]
```diff
@@ -336,6 +382,65 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
         lstq = np.sqrt(s_assoc * s_cls)
         return lstq, s_assoc
 
+    def get_pat(self) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:385]: return Tuple should have 3 arrays.

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 2]
```diff
@@ -336,6 +382,65 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
...
+        """
+        # first calculate for all classes
+        sq_all = self.pan_iou.astype(np.double) / np.maximum(self.pan_tp.astype(np.double), self.eps)
+        rq_all = self.pan_tp.astype(np.double) / np.maximum(
+            self.pan_tp.astype(np.double) + 0.5 * self.pan_fp.astype(np.double) + 0.5 * self.pan_fn.astype(np.double),
+            self.eps)
+        pq_all = sq_all * rq_all
+
+        # then do the REAL mean (no ignored classes)
+        SQ = sq_all[self.include].mean()
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:401]: use low-case variables to be consistent

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 3]
```diff
@@ -186,6 +224,14 @@ def add_batch_panoptic(self,
                 x_inst_row[0] = x_inst_row[0][gt_not_in_excl_mask]
                 y_inst_row[0] = y_inst_row[0][gt_not_in_excl_mask]
 
+        #accumulate class-agnostic predictions
+        unique_pred_, counts_pred_ = np.unique(x_inst_row[1][x_inst_row[1] > 0], return_counts=True) 
+        for p_id in unique_pred_[counts_pred_ > self.min_points]:
+                if p_id not in self.instance_preds[scene]:
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:230]: indentation is incorrect

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 2 above)
Comment (lubing-motional) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:401]: remove unused SQ, RQ.

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 4]
```diff
@@ -336,6 +382,65 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
...
+
+        for seq in self.sequences:
+            preds = self.instance_preds[seq]
+            for cl in self.include:
+                cls_gts = self.instance_gts[seq][cl]
+                for gt_id, pr_ids in cls_gts.items():
+                    unique_pr_id, counts_pr_id = np.unique(pr_ids, return_counts=True)
+
+                    track_length = len(pr_ids)
+                    unique_pr_id, counts_pr_id = unique_pr_id[unique_pr_id!=1], counts_pr_id[unique_pr_id!=1]
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:416]: could you add more inline comments for lines like this, why need to remove pr_id == 1, I feel it is hard to decode all the logic by just reading the code.

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 5]
```diff
@@ -129,6 +133,38 @@ def get_panoptic_track_stats(self,
...
+            pred_areas_ = np.array([counts_pred_[id2idx_pred_[p_id]] for p_id in pred_labels_])
+            intersections_ = counts_combo_
+            unions_ = gt_areas_ + pred_areas_ - intersections_
+            ious_agnostic = intersections_.astype(np.float32) / unions_.astype(np.float32)
+            tp_indexes_agnostic = ious_agnostic > 0.5
+            matched_gt_ = np.array([False] * len(id2idx_gt_))
+            matched_gt_[[id2idx_gt_[g_id] for g_id in gt_labels_[tp_indexes_agnostic]]] = True
+
+            for idx, value in enumerate(tp_indexes_agnostic):
+                if value == True:
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:153]: use `if value:`

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 6]
```diff
@@ -129,6 +133,38 @@ def get_panoptic_track_stats(self,
...
+                    g_label = gt_labels_[idx]  
+                    p_label = pred_labels_[idx]
+                    if g_label not in self.instance_gts[scene][cl]:
+                        self.instance_gts[scene][cl][g_label] = [p_label,]
+                    else:
+                        self.instance_gts[scene][cl][g_label].append(p_label)
+
+            for g_label in unique_gt_:
+                if matched_gt_[id2idx_gt_[g_label]] == False:
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:162]: use `if not matched_gt_[id2idx_gt_[g_label]]`

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 7]
```diff
@@ -129,6 +133,38 @@ def get_panoptic_track_stats(self,
...
+                    p_label = pred_labels_[idx]
+                    if g_label not in self.instance_gts[scene][cl]:
+                        self.instance_gts[scene][cl][g_label] = [p_label,]
+                    else:
+                        self.instance_gts[scene][cl][g_label].append(p_label)
+
+            for g_label in unique_gt_:
+                if matched_gt_[id2idx_gt_[g_label]] == False: 
+                   if g_label not in self.instance_gts[scene][cl]:
+                        self.instance_gts[scene][cl][g_label] = [1,]
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:164]: could you double check the indentations, here has 5 spaces, i see many places have wrong indentations.

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 8]
```diff
@@ -129,6 +133,38 @@ def get_panoptic_track_stats(self,
...
+            id2idx_gt_ = {inst_id: idx for idx, inst_id in enumerate(unique_gt_)}
+            unique_pred_, counts_pred_ = np.unique(x_inst_row[x_inst_row > 0], return_counts=True)
+            id2idx_pred_ = {inst_id: idx for idx, inst_id in enumerate(unique_pred_)}
+            gt_labels_ = unique_combo_ // self.offset
+            pred_labels_ = unique_combo_ % self.offset
+            gt_areas_ = np.array([counts_gt_[id2idx_gt_[g_id]] for g_id in gt_labels_])
+            pred_areas_ = np.array([counts_pred_[id2idx_pred_[p_id]] for p_id in pred_labels_])
+            intersections_ = counts_combo_
+            unions_ = gt_areas_ + pred_areas_ - intersections_
+            ious_agnostic = intersections_.astype(np.float32) / unions_.astype(np.float32)
```
Comment (holger-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:147]: Could unions be 0?

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 9]
```diff
@@ -186,6 +224,14 @@ def add_batch_panoptic(self,
                 x_inst_row[0] = x_inst_row[0][gt_not_in_excl_mask]
                 y_inst_row[0] = y_inst_row[0][gt_not_in_excl_mask]
 
+        #accumulate class-agnostic predictions
```
Comment (holger-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:227]: Format to: # Accumulate

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 4 above)
Comment (holger-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:416]: spacing around !=

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 10]
```diff
@@ -336,6 +382,65 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
...
+                    unique_pr_id, counts_pr_id = unique_pr_id[unique_pr_id!=1], counts_pr_id[unique_pr_id!=1] 
+                    fp_pr_id = []
+                    for idx, uid in enumerate(unique_pr_id):
+                        if uid in preds:
+                            fp_pr_id.append(preds[uid] - counts_pr_id[idx])
+                        else:
+                            fp_pr_id.append(0)
+
+                    fp_pr_id = np.array(fp_pr_id)
+                    gt_id_aq = (np.sum(counts_pr_id**2/np.double(track_length + fp_pr_id)))/np.double(track_length)
```
Comment (holger-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:425]: Let's add more spacing around operators

**Review by holger-motional — APPROVED** [review]

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 11]
```diff
@@ -129,6 +133,38 @@ def get_panoptic_track_stats(self,
             unique_combo_, counts_combo_ = np.unique(offset_combo_, return_counts=True)
             self.update_dict_stat(cl_intersects, unique_combo_, counts_combo_)
 
+            # Computation for PAT score
+            unique_gt_, counts_gt_ = np.unique(y_inst_in_cl[y_inst_in_cl > 0], return_counts=True)
+            id2idx_gt_ = {inst_id: idx for idx, inst_id in enumerate(unique_gt_)}
+            unique_pred_, counts_pred_ = np.unique(x_inst_row[x_inst_row > 0], return_counts=True)
+            id2idx_pred_ = {inst_id: idx for idx, inst_id in enumerate(unique_pred_)}
+            gt_labels_ = unique_combo_ // self.offset
+            pred_labels_ = unique_combo_ % self.offset
```
Comment (whyekit-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:142]: Perhaps some comments on what these two lines are doing?:
```
gt_labels_ = unique_combo_ // self.offset
pred_labels_ = unique_combo_ % self.offset
```

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 12]
```diff
@@ -160,6 +196,8 @@ def add_batch_panoptic(self,
             self.gts[scene] = [{} for _ in range(self.n_classes)]
             self.intersects[scene] = [{} for _ in range(self.n_classes)]
             self.intersects_ovr[scene] = [{} for _ in range(self.n_classes)]
+            self.instance_preds[scene] = {}
+            self.instance_gts[scene] = [{} for _ in range(self.n_classes)]
         # Make sure instance IDs are non-zeros. Otherwise, they will be ignored. Note in nuScenes-panoptic,
```
Comment (whyekit-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:201]: `nuScenes-panoptic` --> `Panoptic nuScenes`?

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 8 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:147]: No. gt_areas_ can never be 0.

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 1 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py]: done

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 2 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py]: done

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 3 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py]: corrected

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 5 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py]: done

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 6 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py]: done

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 7 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:174]: done

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 9 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py]: done

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 10 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py]: done

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 12 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py]: Corrected it

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 11 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:145]: Added the comment. Let me know if it is intuitive.

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 13]
```diff
@@ -336,6 +392,82 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
...
+            preds = self.instance_preds[seq]
+            for cl in self.include:
+                cls_gts = self.instance_gts[seq][cl]
+                for gt_id, pr_ids in cls_gts.items():
+                    unique_pr_id, counts_pr_id = np.unique(pr_ids, return_counts=True)
+
+                    track_length = len(pr_ids)
+                    # void/stuff have instance value 1 due to the +1 in ln205 as well as unmatched gt is denoted by 1 
+                    # Thus we remove 1 from the prediction id list  
+                    unique_pr_id, counts_pr_id = unique_pr_id[unique_pr_id != 1], counts_pr_id[unique_pr_id != 1]
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:426]: tks for the inline comment, sigh, this looks like a hack on top of another hack..

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 14]
```diff
@@ -336,6 +392,82 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
...
+                    # void/stuff have instance value 1 due to the +1 in ln205 as well as unmatched gt is denoted by 1 
+                    # Thus we remove 1 from the prediction id list  
+                    unique_pr_id, counts_pr_id = unique_pr_id[unique_pr_id != 1], counts_pr_id[unique_pr_id != 1] 
+                    fp_pr_id = []
+
+                    # Computes the total false positve for each prediction id:
+                    #     preds[uid]: TPA + FPA (class-agnostic)
+                    #     counts_pr_id[idx]: TPA (class-agnostic)
+                    # If prediction id is not in preds it means it has number of points < self.min_points.
+                    # Similar to PQ computation we consider pred with number of points < self.min_points with IoU overlap greater than 0.5
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:433]: change line if it exceeds 120 line width

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 15]
```diff
@@ -336,6 +392,82 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
...
+                    gt_id_is = 1.0
+
+                    if track_length > 1:
+                        # Compute the ID switch component 
+                        s_id = -1
+                        ids = 0
+                        # Total possible id switches
+                        total_ids = track_length - 1
+                        # Gt tracks with no corresponding prediction match are assigned 1.
+                        # We consider an id switch occurs if previous predicted id and the current one don't match for the given gt track
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:454]: exceed 120 line width

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 16]
```diff
@@ -336,6 +392,82 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
...
+                for gt_id, pr_ids in cls_gts.items():
+                    unique_pr_id, counts_pr_id = np.unique(pr_ids, return_counts=True)
+
+                    track_length = len(pr_ids)
+                    # void/stuff have instance value 1 due to the +1 in ln205 as well as unmatched gt is denoted by 1 
+                    # Thus we remove 1 from the prediction id list  
+                    unique_pr_id, counts_pr_id = unique_pr_id[unique_pr_id != 1], counts_pr_id[unique_pr_id != 1] 
+                    fp_pr_id = []
+
+                    # Computes the total false positve for each prediction id:
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:429]: typo: positive

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 17]
```diff
@@ -336,6 +392,82 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
...
+        rq_all = self.pan_tp.astype(np.double) / np.maximum(
+            self.pan_tp.astype(np.double) + 0.5 * self.pan_fp.astype(np.double) + 0.5 * self.pan_fn.astype(np.double),
+            self.eps)
+        pq_all = sq_all * rq_all
+
+        # Then do the REAL mean (no ignored classes)
+        pq = pq_all[self.include].mean()
+
+        accumulate_tq = 0.0
+        accumlate_norm = 0
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:414]: typo: accumulate_norm

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 18]
```diff
@@ -336,6 +392,82 @@ def get_lstq(self) -> Tuple[np.ndarray, np.ndarray]:
...
+                        for pr_id in pr_ids:
+                            if s_id != -1:
+                                if pr_id != s_id or s_id == 1: 
+                                    ids += 1
+                            s_id = pr_id
+                        gt_id_is = 1-(ids/np.double(total_ids))     
+                    # Accumulate TQ over all the possible unique gt instances
+                    accumulate_tq += np.sqrt(gt_id_aq * gt_id_is)
+                    # Count the total number of unique gt instances
+                    accumlate_norm +=1
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:465]: minor: space after +=, similarly for line 461.

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py [hunk 19]
```diff
@@ -129,6 +133,48 @@ def get_panoptic_track_stats(self,
             unique_combo_, counts_combo_ = np.unique(offset_combo_, return_counts=True)
             self.update_dict_stat(cl_intersects, unique_combo_, counts_combo_)
 
+            # Computation for PAT score
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:136]: could you help to correct the return typing to `Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, Dict[int, int], Dict[int, int], np.ndarray]`  in lines 83-84. tks.

File: python-sdk/nuscenes/eval/panoptic/evaluate.py [hunk 20]
```diff
@@ -240,12 +242,16 @@ def evaluate_tracking(self) -> Dict[str, Any]:
                     break
                 cur_token = cur_sample['next']
 
+        pat, mean_pq, mean_tq = self.evaluator['tracking'].get_pat()
         mean_ptq, class_all_ptq, mean_sptq, class_all_sptq = self.evaluator['tracking'].get_ptq()
         mean_iou, class_all_iou = self.evaluator['tracking'].getSemIoU()
         lstq, s_assoc = self.evaluator['tracking'].get_lstq()
         mean_motsa, mean_s_motsa, mean_motsp = self.evaluator['tracking'].get_motsa()
 
-        results = self.wrap_result_mopt(mean_ptq=mean_ptq,
```
Comment (lubing-motional) [inline python-sdk/nuscenes/eval/panoptic/evaluate.py:248]: update the function doc-string part at line 203-210 to add the added PAT, PQ, TQ fields

**Review by lubing-motional — APPROVED** [review]
Took a second pass, clearer that previous round, even though it's still hard for me to follow the logic for all the details due to the implementation complexity, but  think it's hard to improve that without massive changes. So I'd proceed to approve it due to the time being.

**Comment by lubing-motional** [comment]
@mohan1914 Pls promote the version to 1.1.8 for following 2 places
(1) https://github.com/nutonomy/nuscenes-devkit/blob/master/setup/setup.py#L42
(2) https://github.com/nutonomy/nuscenes-devkit/blob/master/README.md,   Add following line in the ChangeLog 
"Sep 17, 2021: Devkit v1.1.8: Add  PAT metric to Panoptic nuScenes."

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 13 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:426]: Yup

File: python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py (hunk 15 above)
Comment (mohan1914) [reply python-sdk/nuscenes/eval/panoptic/panoptic_track_evaluator.py:454]: done

**Pull request merged**